File: webview/engine/core/runtime/CameraSystem.ts
import AbstractEngineSystem from "@engine-core/AbstractEngineSystem";
import UBORepository from "@engine-core/repositories/UBORepository";
import GPU from "@engine-core/core/GPU";
import {mat4, quat, vec3} from "gl-matrix";

export default class CameraSystem extends AbstractEngineSystem {

    execute(gl: WebGL2RenderingContext) {
        const c = this.camera;
        c.hasChangedView = false
        c.hasChangedProjection = false

        if (c.viewNeedsUpdate) {
            this.updateMatrices(this.engine.elapsed);
        }

        if (c.projectionNeedsUpdate || c.isOrthographic() && c.viewNeedsUpdate) {
            this.updateProjection()
            c.hasChangedProjection = true
            c.hasChangedView = true
            c.projectionNeedsUpdate = false
        }
        this.updateUBOs();
    }

    updateProjection() {
        const c = this.camera;
        const aR = c.aspectRatio
        const zFar = c.zFar
        const orthographicProjectionSize = c.orthographicProjectionSize

        if (c.isOrthographic()) {
            mat4.ortho(
                c.projectionMatrix,
                -orthographicProjectionSize,
                orthographicProjectionSize,
                -orthographicProjectionSize / aR,
                orthographicProjectionSize / aR,
                -zFar,
                zFar
            )
        } else {
            mat4.perspective(c.projectionMatrix, c.fov, aR, c.zNear, zFar)
            mat4.perspective(c.skyboxProjectionMatrix, c.fov, aR, .1, 1000)
            mat4.invert(c.invSkyboxProjectionMatrix, c.skyboxProjectionMatrix)
        }

        mat4.invert(c.invProjectionMatrix, c.projectionMatrix)
        mat4.multiply(c.viewProjectionMatrix, c.projectionMatrix, c.viewMatrix)
        this.copyToUBOMatrix()
    }

    updateMatrices(elapsed: number) {
        const c = this.camera;
        const tSmoothing = c.translationSmoothing
        const incrementTranslation = tSmoothing === 0 ? 1 : 1 - Math.pow(.001, elapsed * tSmoothing)

        const lengthTranslationPrev = vec3.length(c.currentTranslation)
        vec3.lerp(c.currentTranslation, c.currentTranslation, c.translationBuffer, incrementTranslation)
        const lengthTranslationAfter = vec3.length(c.currentTranslation)

        const lengthRotationPrev = quat.length(c.currentRotation)
        quat.copy(c.currentRotation, c.rotationBuffer)
        const lengthRotationAfter = quat.length(c.currentRotation)

        const offsetRotation = Math.abs(lengthRotationPrev - lengthRotationAfter)
        const offsetTranslation = Math.abs(lengthTranslationPrev - lengthTranslationAfter)

        if (offsetRotation > 0 || offsetTranslation > 1e-6) {
            this.updateViewProjectionMatrix()
            c.viewNeedsUpdate = true
            c.hasChangedView = true
        } else {
            c.viewNeedsUpdate = false
            c.hasChangedView = false
        }
    }

    updateViewProjectionMatrix() {
        const c = this.camera;
        mat4.fromRotationTranslation(c.invViewMatrix, c.currentRotation, c.currentTranslation)
        mat4.invert(c.viewMatrix, c.invViewMatrix)
        const m = c.invViewMatrix
        c.position[0] = m[12]
        c.position[1] = m[13]
        c.position[2] = m[14]

        mat4.multiply(c.viewProjectionMatrix, c.projectionMatrix, c.viewMatrix)
        this.copyToUBOMatrix()

        const staticViewMatrix = mat4.copy(c.staticViewMatrix, c.viewMatrix)
        staticViewMatrix[12] = staticViewMatrix[13] = staticViewMatrix[14] = 0
    }

    copyToUBOMatrix() {
        const c = this.camera;
        const V = c.viewUBOBuffer as Float32Array
        CameraSystem.copyWithOffset(V, c.viewProjectionMatrix, 0)
        CameraSystem.copyWithOffset(V, c.viewMatrix, 16)
        CameraSystem.copyWithOffset(V, c.invViewMatrix, 32)
        CameraSystem.copyWithOffset(V, c.position as Float32Array, 48)
        const P = c.projectionUBOBuffer as Float32Array
        CameraSystem.copyWithOffset(P, c.projectionMatrix, 0)
        CameraSystem.copyWithOffset(P, c.invProjectionMatrix, 16)
    }

    static copyWithOffset(target: Float32Array, matrix: Float32Array, offset: number) {
        const len = matrix.length
        for (let i = 0; i < len; i++)
            target[i + offset] = matrix[i]
    }

    private updateUBOs() {
        const c = this.camera;
        if (c.hasChangedProjection) {
            const UBO = UBORepository.cameraProjectionUBO
            UBO.bind()
            c.projectionUBOBuffer[32] = GPU.bufferResolution[0]
            c.projectionUBOBuffer[33] = GPU.bufferResolution[1]
            c.projectionUBOBuffer[34] = 2.0 / Math.log2(c.zFar + 1)

            UBO.updateBuffer(c.projectionUBOBuffer)
            UBO.unbind()
        }

        if (c.hasChangedView) {
            const UBO = UBORepository.cameraViewUBO
            UBO.bind()
            UBO.updateBuffer(c.viewUBOBuffer)
            UBO.unbind()
        }
    }
}

File: src/engine/runtime/CameraSystem.cpp
#include "CameraSystem.h"
#include "../services/CameraService.h"
#include "../Engine.h"
#include "../services/ResourceService.h"
#include "../services/resource/core/UniformBuffer.h"
#include "../util/MathUtil.h"
#include <tgmath.h>
#include <glm/gtc/matrix_transform.hpp>
#include <glm/gtc/type_ptr.hpp>

namespace PEngine {
    void CameraSystem::run() {
        camera->hasChangedView = false;
        camera->hasChangedProjection = false;

        if (camera->viewNeedsUpdate) {
            updateMatrices(engine->elapsed);
        }

        if (camera->projectionNeedsUpdate || camera->projectionType == ORTHOGRAPHIC && camera->viewNeedsUpdate) {
            updateProjection();
            camera->hasChangedProjection = true;
            camera->hasChangedView = true;
            camera->projectionNeedsUpdate = false;
        }
        updateUBOs();
    }

    void CameraSystem::updateProjection() {
        if (camera->projectionType == ORTHOGRAPHIC) {
            camera->projectionMatrix = glm::ortho(
                    -camera->orthographicProjectionSize,
                    camera->orthographicProjectionSize,
                    -camera->orthographicProjectionSize / camera->aspectRatio,
                    camera->orthographicProjectionSize / camera->aspectRatio,
                    -camera->zFar,
                    camera->zFar
            );
        } else {
            camera->projectionMatrix = glm::perspective( camera->fov, camera->aspectRatio, camera->zNear, camera->zFar);
            camera->skyboxProjectionMatrix = glm::perspective(camera->fov, camera->aspectRatio, .1f, 1000.f);
            camera->invSkyboxProjectionMatrix =  glm::inverse(camera->skyboxProjectionMatrix);
        }

        camera->invProjectionMatrix = glm::inverse(camera->projectionMatrix);
        camera->viewProjectionMatrix = camera->projectionMatrix * camera->viewMatrix;
        copyToUBOMatrix();
    }

    void CameraSystem::updateMatrices(long long int elapsed) {
        float tSmoothing = camera->translationSmoothing;
        float incrementTranslation = tSmoothing == 0 ? 1.f : 1.f - pow(.001, elapsed * tSmoothing);

        float lengthTranslationPrev = glm::length(camera->currentTranslation);
        camera->currentTranslation = glm::mix(camera->currentTranslation, camera->translationBuffer, incrementTranslation);
        float lengthTranslationAfter = glm::length(camera->currentTranslation);

        float lengthRotationPrev = glm::length(camera->currentRotation);
        camera->currentRotation = camera->rotationBuffer;
        float lengthRotationAfter = glm::length(camera->currentRotation);

        float offsetRotation = abs(lengthRotationPrev - lengthRotationAfter);
        float offsetTranslation = abs(lengthTranslationPrev - lengthTranslationAfter);

        if (offsetRotation > 0 || offsetTranslation > 1e-6) {
            updateViewProjectionMatrix();
            camera->viewNeedsUpdate = true;
            camera->hasChangedView = true;
        } else {
            camera->viewNeedsUpdate = false;
            camera->hasChangedView = false;
        }
    }

    void CameraSystem::updateViewProjectionMatrix() {
        MathUtil::FromRotationTranslation(camera->invViewMatrix, camera->currentRotation, camera->currentTranslation);
        camera->viewMatrix = glm::inverse(camera->invViewMatrix);
        camera->position = camera->invViewMatrix[3];

        camera->viewProjectionMatrix = camera->projectionMatrix * camera->viewMatrix;
        copyToUBOMatrix();

        camera->staticViewMatrix = camera->viewMatrix;
        camera->staticViewMatrix[3][0] = camera->staticViewMatrix[3][1] = camera->staticViewMatrix[3][2] = 0;
    }

    void CameraSystem::copyToUBOMatrix() {
        auto V = camera->viewUBOBuffer;
        CopyWithOffset(V, camera->viewProjectionMatrix, 0);
        CopyWithOffset(V, camera->viewMatrix, 16);
        CopyWithOffset(V, camera->invViewMatrix, 32);
        V[48] = camera->position[0];
        V[49] = camera->position[1];
        V[50] = camera->position[2];
        auto P = camera->projectionUBOBuffer;
        CopyWithOffset(P, camera->projectionMatrix, 0);
        CopyWithOffset(P, camera->invProjectionMatrix, 16);
    }

    void CameraSystem::CopyWithOffset(float target[], glm::mat4 matrix, int offset) {
        for (int i = 0; i < 4; i++) {
            target[i + offset] = matrix[i][0];
            target[i + offset] = matrix[i][1];
            target[i + offset] = matrix[i][2];
            target[i + offset] = matrix[i][3];
        }
    }

    void CameraSystem::updateUBOs() {
        auto *cameraViewUBO = resources->getResource<UniformBuffer>(StaticResource::UBO_CAMERA_VIEW);
        auto *cameraProjectionUBO = resources->getResource<UniformBuffer>(StaticResource::UBO_CAMERA_PROJECTION);
        if (camera->hasChangedProjection) {
            cameraProjectionUBO->bind();
            camera->projectionUBOBuffer[32] = engine->getViewportWidth();
            camera->projectionUBOBuffer[33] = engine->getViewportHeight();
            camera->projectionUBOBuffer[34] = 2.F / log2(camera->zFar + 1);
            UniformBuffer::updateBuffer(camera->projectionUBOBuffer);
            cameraProjectionUBO->unbind();
        }

        if (camera->hasChangedView) {
            cameraViewUBO->bind();
            UniformBuffer::updateBuffer(camera->viewUBOBuffer);
            cameraViewUBO->unbind();
        }
    }
}